internal/indexer/postgres/affectedmanifest.go

```diff
@@ -0,0 +1,269 @@
+					repo_id = $3
+				END
+			);
+		`
+	)
+	log := zerolog.Ctx(ctx).
+		With().
+		Str("component", "internal/indexer/postgres/affectedManifests").
+		Logger()
+
+	// confirm the incoming vuln can be
+	// resolved into a prototype index record
+	pr, err := protoRecord(ctx, pool, v)
+	switch {
+	case err == nil:
+		// break out
+	case errors.Is(err, ErrNotIndexed):
+		log.Warn().Str("vulnID", v.ID).Msg("vuln is comprised of repo or distro not indexed by any scanner yet")
+		return nil, nil
+	default:
+		return nil, err
+	}
+
+	// collect all packages which may be affected
+	// by the vulnerability in question.
+	pkgsToFilter := []claircore.Package{}
+	rows, err := pool.Query(ctx, selectPackages, v.Package.Name)
+	if err != nil {
+		rows.Close()
+		if errors.Is(err, pgx.ErrNoRows) {
+			return []claircore.Digest{}, nil
+		}
+		return nil, fmt.Errorf("failed to query packages associated with vulnerablility %+v: %v", v, err)
+	}
+	for rows.Next() {
+		var pkg claircore.Package
+		var id int64
+		err := rows.Scan(
+			&id,
+			&pkg.Name,
+			&pkg.Version,
+			&pkg.Kind,
+		)
+		if err != nil {
+			rows.Close()
+			return nil, fmt.Errorf("failed to scan package: %v", err)
+		}
+		idStr := strconv.FormatInt(id, 10)
+		pkg.ID = idStr
+		pkgsToFilter = append(pkgsToFilter, pkg)
+	}
+	log.Debug().Int("count", len(pkgsToFilter)).Msg("packages to filter")
+
+	// for each package discovered create an index record
+	// and determine if any in-tree matcher finds the record vulnerable
+	var filteredRecords []claircore.IndexRecord
+	om := omnimatcher.New(nil)
+	for _, pkg := range pkgsToFilter {
+		pr.Package = &pkg
+		if om.Vulnerable(&pr, &v) {
+			p := pkg // make a copy, or else you'll get a stale reference later
+			filteredRecords = append(filteredRecords, claircore.IndexRecord{
+				Package:      &p,
+				Distribution: pr.Distribution,
+				Repository:   pr.Repository,
+			})
+		}
+	}
+	log.Debug().Int("count", len(filteredRecords)).Msg("vulnerable indexrecords")
+
+	// query the manifest index for manifests containing
+	// the vulnerable indexrecords and create a set
+	// containing each unique manifest
+	set := map[string]claircore.Digest{}
+	for _, record := range filteredRecords {
+		v, err := toValues(record)
+		if err != nil {
+			return nil, fmt.Errorf("failed to resolve record %+v to sql values for query: %v", record, err)
+		}
+
+		rows, err := pool.Query(ctx,
+			selectAffected,
+			record.Package.ID,
+			v[2],
+			v[3],
+		)
+		if err != nil {
+			rows.Close()
+			if err == pgx.ErrNoRows {
+				continue
+			}
+			return nil, fmt.Errorf("failed to query the manifest index: %v", err)
+		}
+		for rows.Next() {
+			var hash claircore.Digest
+			err := rows.Scan(&hash)
+			if err != nil {
+				rows.Close()
+				return nil, fmt.Errorf("failed scanning manifest hash into digest: %v", err)
+			}
+			set[hash.String()] = hash
+		}
+	}
+
+	n := len(set)
+	out := make([]claircore.Digest, 0, n)
+	for _, v := range set {
+		out = append(out, v)
+	}
```

Comment:
This could avoid an extra copy by making `set` a `map[string]struct{}`, then doing an existence check and an append if that fails here. Otherwise, this is copying every Digest into a map, then copying the last writes again into a slice.